scio-extra/src/main/scala/com/spotify/scio/extra/annoy/package.scala

```diff
@@ -0,0 +1,214 @@
+ *    }
+ *  }
+ *
+ * }}}
+ */
+package object annoy {
+
+  sealed trait ScioAnnoyMetric
+  case object Angular extends ScioAnnoyMetric
+  case object Euclidean extends ScioAnnoyMetric
+
+  val logger = LoggerFactory.getLogger(this.getClass)
+
+  /**
+   *  Wraps annoy-java
+   */
+  class AnnoyReader(metric: ScioAnnoyMetric, dim: Int, nTrees: Int, path: String) {
+
+    import com.spotify.annoy._
+
+    val indexType = metric match {
+      case Angular => IndexType.ANGULAR
+      case Euclidean => IndexType.EUCLIDEAN
+    }
+
+    val index = new ANNIndex(dim, path, indexType)
+
+    def getItemVector(i: Int): Array[Float] = index.getItemVector(i)
+    def getItemVector(i: Int, v: Array[Float]): Unit = index.getItemVector(i, v)
+    def getNearest(v: Array[Float], maxNumResults: Int): Array[Int] = {
+      index.getNearest(v, maxNumResults).asScala.map(Int.box(_).toInt).toArray[Int]
+    }
+
+  }
+
+  /** Enhanced version of [[ScioContext]] with Annoy methods */
+  implicit class AnnoyScioContext(val self: ScioContext) extends AnyVal {
+    /**
+      * Create a SideInput of `AnnoyIndex` from an [[AnnoyUri]] base path, to be used with
+      * [[com.spotify.scio.values.SCollection.withSideInputs SCollection.withSideInputs]]
+      */
+    def annoySideInput(metric: ScioAnnoyMetric, dim: Int, nTrees: Int, path: String)
+    : SideInput[AnnoyReader] = {
+      val uri = AnnoyUri(path, self.options)
+      val view = self.parallelize(Seq(uri)).applyInternal(View.asSingleton())
+      new AnnoySideInput(view, metric, dim, nTrees)
+    }
+  }
+
+  implicit class AnnoyPairSCollection(val self: SCollection[(Int, Array[Float])]) {
+    /**
+      * Write the key-value pairs of this SCollection as an Annoy file to a specific location.
+      *
+      * @return A singleton SCollection containing the [[AnnoyUri]] of the saved files
+      */
+    def asAnnoy(metric: ScioAnnoyMetric, dim: Int, nTrees: Int, path: String)
+    : SCollection[AnnoyUri] = {
+      val uri = AnnoyUri(path, self.context.options)
+      require(!uri.exists, s"Annoy URI ${uri.path} already exists")
+
+      self.transform { in =>
+        in.groupBy(_ => ())
+          .map { case(_, xs) =>
+            logger.info(s"Saving as Annoy: $uri")
+            val annoyIndex = new AnnoyWriter(metric, dim, nTrees)
+            val it = xs.iterator
+            while (it.hasNext) {
+              val kv = it.next()
+              annoyIndex.addItem(kv._1, kv._2)
+            }
+            uri.saveAndClose(annoyIndex)
+            uri
+          }
+      }
+    }
+
+    /**
+      * Write the key-value pairs of this SCollection as an Annoy file to a temporary location.
+      *
+      *   @return A Singleton SCollection containing the [[AnnoyUri]] of the saved files
+      */
+    def asAnnoy(metric: ScioAnnoyMetric, dim: Int, nTrees: Int)
+    : SCollection[AnnoyUri] = {
+      val uuid = UUID.randomUUID()
+      val path = self.context.options.getTempLocation + s"/annoy-build-$uuid"
```

Comment:
probably the whole thing could be in `s""` also is there a need to check if `getTempLocation` ends with `/`?